sdlib/d/gc/tcache.d

```diff
@@ -151,34 +163,33 @@ public:
 	 * 
 	 * See also: https://dlang.org/spec/arrays.html#capacity-reserve
 	 */
-	bool getAppendablePageDescriptor(const void[] slice,
-	                                 ref PageDescriptor pd) {
+	bool getAppendablePageDescriptor(const void[] slice, ref PageDescriptor pd,
+	                                 ref CapacityInfo info) {
 		pd = maybeGetPageDescriptor(slice.ptr);
 		if (pd.extent is null) {
 			return false;
 		}
 
-		// Appendable slabs are not supported.
-		if (pd.isSlab()) {
-			return false;
-		}
+		info = getCapacityInfo(pd, cast(void*) slice.ptr);
 
 		// Slice must not end before valid data ends, or capacity is zero:
-		auto startIndex = slice.ptr - pd.extent.address;
+		auto startIndex = slice.ptr - info.address;
 		auto stopIndex = startIndex + slice.length;
 
 		// If the slice end doesn't match the used capacity, not appendable.
-		return stopIndex == pd.extent.usedCapacity;
+		// A zero-length slice at the start of an empty alloc is also not appendable.
+		return (stopIndex == info.usedCapacity) && (stopIndex > 0);
```

Comment:
This can also go on its own with it own test. this is a behavior change that affect large allocations anyways (and should, the problem exist for large allocations too! It would also make it very evident that this is tested rather than trying to hunt down some assert in a sea of green and red.

Reply:
And this change plus its test would probably be merged already if they had been int heir own PR, because you wouldn't have to explains this, nor me have to comb through 100s of line of test to figure if this is actually the case.